hotham/src/shaders/pbr.frag

```diff
@@ -25,277 +47,257 @@ layout(set = 0, binding = 5) uniform samplerCube cubeTextures[];
-	float reflectance90 = clamp(reflectance * 25.0, 0.0, 1.0);
-	vec3 specularEnvironmentR0 = specularColor.rgb;
-	vec3 specularEnvironmentR90 = vec3(1.0, 1.0, 1.0) * reflectance90;
+	// Get NdotV
+	float NdotV = clamp(abs(dot(n, v)), 0., 1.0);
 
-	vec3 n = (material.normalTextureID == NO_TEXTURE) ? normalize(inNormal) : getNormal(material.normalTextureID);
-	vec3 v = normalize(sceneData.cameraPosition[gl_ViewIndex].xyz - inGlobalPos);    // Vector from surface point to camera
-	vec3 l = normalize(sceneData.lightDirection.xyz);     // Vector from surface point to light
-	vec3 h = normalize(l+v);                        // Half vector between both l and v
 	vec3 reflection = -normalize(reflect(v, n));
 
-	// TODO: Is this correct?
-	reflection.y *= -1.;
-
-	float NdotL = clamp(dot(n, l), 0.001, 1.0);
-	float NdotV = clamp(abs(dot(n, v)), 0.001, 1.0);
-	float NdotH = clamp(dot(n, h), 0.0, 1.0);
-	float LdotH = clamp(dot(l, h), 0.0, 1.0);
-	float VdotH = clamp(dot(v, h), 0.0, 1.0);
-
-	PBRInfo pbrInputs = PBRInfo(
-		NdotL,
-		NdotV,
-		NdotH,
-		LdotH,
-		VdotH,
-		perceptualRoughness,
-		metallic,
-		specularEnvironmentR0,
-		specularEnvironmentR90,
-		alphaRoughness,
+	// Collect all the material info.
+	MaterialInfo materialInfo = MaterialInfo(
+		f0,
 		diffuseColor,
-		specularColor
+		specularColor,
+		alphaRoughness,
+		perceptualRoughness
 	);
 
-	// Calculate the shading terms for the microfacet specular shading model
-	vec3 F = specularReflection(pbrInputs);
-	float G = geometricOcclusion(pbrInputs);
-	float D = microfacetDistribution(pbrInputs);
-
-	// Calculation of analytical lighting contribution
-	vec3 diffuseContrib = (1.0 - F) * diffuse(pbrInputs);
-	vec3 specContrib = F * G * D / (4.0 * NdotL * NdotV);
-
-	// Obtain final intensity as reflectance (BRDF) scaled by the energy of the light (cosine law)
-	vec3 color = NdotL * (diffuseContrib + specContrib);
-
 	// Calculate lighting contribution from image based lighting source (IBL), scaled by a scene data parameter.
-	color += getIBLContribution(pbrInputs, n, reflection) * sceneData.params.x;
+	vec3 color;
+	if (sceneData.params.x > 0.) {
+		color = getIBLContribution(materialInfo, n, reflection, NdotV) * sceneData.params.x;
+	} else {
+		color = vec3(0.);
+	}
 
-	// Apply optional PBR terms for additional (optional) shading
-	if (material.occlusionTextureID != NO_TEXTURE) {
+	// Apply ambient occlusion, if present.
+	if (material.occlusionTextureID != NOT_PRESENT) {
 		// Occlusion is stored in the 'g' channel as per:
 		// https://github.com/ARM-software/astc-encoder/blob/main/Docs/Encoding.md#encoding-1-4-component-data
 		float ao = texture(textures[material.occlusionTextureID], inUV).g;
 		color = color * ao;
 	}
 
-	if (material.emissiveTextureID != NO_TEXTURE) {
+    // Walk through each light and add its color contribution.
+	// Qualcomm's documentation suggests that loops are undesirable, so we do branches instead.
```

Comment:
Ah, you may be right! I'd like to profile this in an optimisation pass, but now that I re-read the documentation re: unrolling loops it's pretty ambiguous:

https://developer.qualcomm.com/sites/default/files/docs/adreno-gpu/developer-guide/gpu/best_practices_shaders.html?highlight=loop#minimize-usage-of-shader-gprs

The other trick here is that this method doesn't require us to specify a `lights_count` to iterate through, which is kind of neat, but I agree it may be suboptimal.